unit_testing/LightController/LightController_tests.cpp

```diff
@@ -0,0 +1,432 @@
+            },
+            .productionDate = 0
+        }
+    };
+    HAL_otpWrite(&bi);
+
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+
+    /* Minor version does not add up. Produce error.  */
+    EXPECT_TRUE(bsGetField(BS_VERSION_ERROR_Msk));
+
+    /* This should force a print on the USB bus */
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+
+    /* Check that it prints the status error code */
+    EXPECT_FLUSH_USB(Contains("0x84000000"));
+    
+
+    /* Reset to correct params */
+    bsClearField(BS_VERSION_ERROR_Msk);
+    bsClearError(BS_VERSION_ERROR_Msk);
+    bi.v2.pcbVersion.minor=1;
+    HAL_otpWrite(&bi);
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    EXPECT_FALSE(bsGetField(BS_VERSION_ERROR_Msk));
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre("\r", "0, 0, 0, 0x0"));
+
+    /* Run with wrong board type */
+    bi.v2.boardType = AC_Board;
+    HAL_otpWrite(&bi);
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    EXPECT_TRUE(bsGetField(BS_VERSION_ERROR_Msk));
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre("\r", "0x84000000"));
+}
+
+
+TEST_F(LightControllerTest, printSerial) 
+{
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    /* Note: usb RX buffer is flushed during the first loop, so a single loop must be done before
+    ** printing anything */
+    LightControllerLoop(bootMsg);
+    writeLightControllerMessage("Serial\n");
+    
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r", 
+        "Boot Unit Test\r", 
+        "Serial Number: 000\r", 
+        "Product Type: LightController\r", 
+        "Sub Product Type: 0\r", 
+        "MCU Family: Unknown 0x  0 Rev 0\r", 
+        "Software Version: (null)\r", 
+        "Compile Date: (null)\r", 
+        "Git SHA: (null)\r", 
+        "PCB Version: 1.1\r"
+    ));
+}
+
+TEST_F(LightControllerTest, printStatus) 
+{
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    /* Note: usb RX buffer is flushed during the first loop, so a single loop must be done before
+    ** printing anything */
+    LightControllerLoop(bootMsg);
+    writeLightControllerMessage("Status\n");
+    
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "Boot Unit Test\r",
+        "Board Status:\r",
+        "The board is operating normally.\r",  
+        "Port 1: On: 0\r",
+        "Port 2: On: 0\r",
+        "Port 3: On: 0\r",
+        "\r",
+        "End of board status. \r"
+    ));
+}
+
+
+/* Note: this test uses some knowledge of internals of LightController.c (e.g. white box), which isn't 
+** perfect */
+TEST_F(LightControllerTest, GpioInit) 
+{
+    auto expectStmNull = [](StmGpio* stm) {
+        EXPECT_EQ(stm->set,    nullptr);
+        EXPECT_EQ(stm->get,    nullptr);
+        EXPECT_EQ(stm->toggle, nullptr);
+    };
+
+    auto expectStmNotNull = [](StmGpio* stm) {
+        EXPECT_NE(stm->set,    nullptr);
+        EXPECT_NE(stm->get,    nullptr);
+        EXPECT_NE(stm->toggle, nullptr);
```

Comment:
I know this is copied from me, but you may want to use ASSERT_NE instead, because if they _are_ null, line 256 will segfault